### Imports and Configuration

In [1]:
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
pd.set_option('display.max_rows', 200)

### Load WSV Dataset

In [77]:
df = pd.read_csv('C:/Users/USERS/MDA_PROJECT/Datasets_Westernization_R_scripts_regression/country_wave_level_means.csv')
df.head(10)

Country_Wave      A001      A002      A003      A004      A005      A006  \
0         ALB3  1.044266  1.946502  2.573057  3.098765  1.245703  2.260246   
1         ALB4  1.047094  1.855992  2.455375  2.956078  1.240647  2.266055   
2         AND5  1.139581  1.592408  1.547358  2.864729  1.606855  3.034000   
3         AND7  1.115538  1.535857  1.421315  2.940120  1.525948  2.965070   
4         ARG1       NaN       NaN       NaN       NaN       NaN       NaN   
5         ARG2  1.101796  1.656971  1.890010  2.951807  1.284137  2.096386   
6         ARG3  1.142725  1.686165  2.027128  2.960711  1.340130  2.059813   
7         ARG4  1.118061  1.715190  1.901195  3.134328  1.316286  1.903302   
8         ARG5  1.111111  1.569277  1.780020  2.934542  1.352705  2.160966   
9         ARG6  1.123541  1.570039  1.870117  2.880906  1.580454  2.451581   

       D057      D059      D060  ...      E026      E027      E028      E033  \
0  2.839045  2.308442  3.041973  ...  2.582375  2.020095  2.961538  5.069490   
1  2.579587  2.470076  3.200000  ...  2.656109  1.989822  2.911215  5.340717   
2  2.552147  3.145473  3.344344  ...  2.505572  1.779899       NaN  4.638858   
3  2.567376  3.332997  3.552448  ...  2.574277  2.054835  2.349650  4.857612   
4       NaN       NaN       NaN  ...  2.748775  2.227924  2.661765  5.261830   
5  2.245262       NaN       NaN  ...  2.840885  2.489209  2.743992  5.457447   
6  2.277445  2.698364  3.023787  ...  2.862002  2.401905  2.790276  5.688073   
7  2.061017  2.800000  3.110390  ...  2.888034  2.503252  2.794599  6.003448   
8  2.217241  2.829811  3.189979  ...  2.768977  2.213447       NaN  5.526087   
9  2.433923  2.930470  3.170683  ...  2.733964  2.224615  2.340625  5.344994   

       E035      E036      E037      E039      E040      B008  
0  4.880041  3.594292  7.062374  3.305414  2.909091  1.570608  
1  5.959391  3.848075  6.403015  2.972945       NaN  1.562570  
2  6.143574  4.884694  6.358641  4.423423  5.715145  1.229525  
3  6.142430  5.062874  6.533400  3.132470  4.575274  1.344546  
4       NaN       NaN       NaN       NaN       NaN       NaN  
5  7.005155  3.913677  4.971698  3.402335  4.823028       NaN  
6  5.955535  5.075173  5.703810  3.900188  5.051838  1.704612  
7  4.899112  5.656000  6.672859  4.171913       NaN  1.710997  
8  5.328542  6.945312  6.088751  4.711157  4.987705  1.306034  
9  4.746680  6.212903  6.115822  4.710010  4.216410  1.431203  

[10 rows x 39 columns]

### Drop Nans

In [3]:
nan_percentage = df.isna().mean() * 100 # Percentage of Nans for question (column)
print(nan_percentage)
nan_rows = df.drop(columns=['Country_Wave']).isna().all(axis=1)
num_nan_rows = nan_rows.sum()
print("There are ",num_nan_rows, "observations with Nans for all variables")
df_cleaned = df.dropna() # take the observation for which we have Nans
df_cleaned

Country_Wave     0.000000
A001             4.713805
A002             4.713805
A003             4.713805
A004             4.713805
A005             4.713805
A006             4.713805
D057             6.397306
D059            11.447811
D060            11.447811
F115             5.723906
F116             4.713805
F117             1.683502
F118             4.713805
F119            11.447811
F120             2.356902
F121             1.346801
F122            14.141414
F123             3.367003
E114            12.457912
E115            12.794613
E116            14.478114
E117            12.794613
E012             4.713805
E015             1.346801
E016             1.346801
E018             1.010101
E023             2.356902
E025             5.387205
E026             6.734007
E027             5.723906
E028            25.589226
E033            10.437710
E035             4.713805
E036             6.734007
E037             4.713805
E039             9.090909
E040            20.202020
B008        

Country_Wave      A001      A002      A003      A004      A005      A006  \
0           ALB3  1.044266  1.946502  2.573057  3.098765  1.245703  2.260246   
3           AND7  1.115538  1.535857  1.421315  2.940120  1.525948  2.965070   
6           ARG3  1.142725  1.686165  2.027128  2.960711  1.340130  2.059813   
10          ARG7  1.091725  1.541376  1.809810  2.807193  1.468405  2.393939   
11          ARM3  1.157551  1.647413  2.089652  2.526127  1.396690  2.179787   
12          ARM6  1.032022  1.764384  1.959032  2.912168  1.373383  1.531392   
14          AUS3  1.120605  1.415196  1.665682  2.489665  1.663376  2.483522   
17          AUS7  1.110372  1.479615  1.646355  2.411832  1.995392  2.897727   
18          AZE3  1.166248  1.733367  1.891566  2.740946  1.566363  1.903242   
19          AZE6  1.074850  1.791417  2.124750  2.926148  1.496008  2.039920   
23          BGD7  1.015000  2.068562  2.191758  3.076329  1.222594  1.067669   
24          BGR3  1.129699  1.726755  2.232422  2.994225  1.647287  2.674805   
26          BIH3  1.032554  1.318258  1.671441  2.594755  1.234950  1.984861   
29          BLR3  1.179759  1.879185  1.982656  2.713445  1.699411  2.394366   
31          BOL7  1.202323  2.366990  1.788527  2.780236  1.252056  1.596415   
33          BRA3  1.080560  1.565294  1.571804  2.523767  1.169002  1.487292   
35          BRA6  1.134007  1.855698  1.813514  2.751864  1.393122  1.595687   
36          BRA7  1.169694  1.794770  1.827626  2.694220  1.439704  1.735613   
41          CAN7  1.246640  1.611249  1.536088  2.387257  1.994276  2.875062   
46          CHL3  1.116162  2.128283  1.721429  3.171779  1.403885  1.869254   
50          CHL7  1.079000  1.701403  1.427711  2.948795  1.674089  2.418090   
59          COL6  1.155423  2.011258  1.664901  3.024585  1.258278  1.592323   
60          COL7  1.205263  1.970395  1.592105  2.782895  1.317763  1.667763   
63          CYP7  1.086259  1.607251  1.821862  2.819106  1.640244  1.851172   
65          CZE3  1.107517  1.726556  2.018437  2.998236  1.705365  3.148381   
69          DEU7  1.129011  1.433159  1.706693  2.191225  1.799463  2.734560   
70          DOM3  1.144578  1.760000  2.032911  2.636132  1.390000  1.674938   
73          ECU6  1.021631  1.968386  1.593672  2.551581  1.169717  1.477537   
74          ECU7  1.126043  2.146566  1.670578  2.745992  1.232247  1.680301   
80          ESP3  1.193388  1.701493  1.879066  3.119730  1.539354  2.305902   
84          EST3  1.243350  1.881256  2.054726  2.916749  1.478607  2.973896   
87          ETH7  1.053659  1.574797  1.856444  2.793703  1.097561  1.073984   
89          FIN3  1.216189  1.463739  1.741007  3.064748  1.678019  2.589377   
94          GEO3  1.067797  1.297957  1.865230  2.681116  1.704580  1.726910   
96          GEO6  1.016639  1.288686  2.016653  2.733555  1.427144  1.188020   
99          GRC7  1.070833  1.502924  1.722781  2.845767  1.358090  1.712615   
101         GTM7  1.135495  1.793849  1.632585  2.495428  1.458853  1.711554   
103         HKG6  1.172345  1.631000  1.794975  2.527694  1.929860  2.770541   
104         HKG7  1.237108  1.751325  1.837603  2.570252  1.907819  2.838242   
105         HRV3  1.172529  1.604690  1.928632  3.010998  1.556024  2.337553   
106         HTI6  1.305093  1.734075  2.303076  2.574823  1.416121  2.030380   
108         HUN3  1.138675  1.887172  1.952234  3.010819  1.657407  2.647149   
112         IDN7  1.023750  1.718310  1.745074  2.499050  1.080963  1.020000   
114         IND3  1.102161  1.870489  2.743500  2.865453  1.198323  1.792640   
133         JPN3  1.103448  1.559501  1.696911  2.190091  1.561214  3.044699   
137         JPN7  1.082774  1.677154  1.637931  2.211464  1.812261  3.255738   
145         KOR3  1.099678  1.581114  1.921711  2.207455  1.512966  2.446964   
149         KOR7  1.114056  1.622490  1.843373  2.354217  1.742169  2.702008   
153         LBY6  1.036671  1.433540  1.836862  2.284668  1.319885  1.037987   
154 

Since we are left with too few observations (105), instead of dropping them we decide to treat the Nans by substituting the missing values by the mean. Also, we check that there are no observations with Nans for all variables.

### Treat Nans

In [4]:
id_column = df['Country_Wave']  # Replace 'Country_Wave' with the name of your ID column
df_numeric = df.drop(columns=['Country_Wave'])  # Drop the non-numeric column for imputation
df_numeric = df_numeric.dropna(how='all') # Although we know there is no such a row/observation
df_mean = df_numeric.fillna(df_numeric.mean()) # Replace NaNs with the column mean
df_mean['Country_Wave'] = id_column  # Add the id_column
df_mean

A001      A002      A003      A004      A005      A006      D057  \
0    1.044266  1.946502  2.573057  3.098765  1.245703  2.260246  2.839045   
1    1.047094  1.855992  2.455375  2.956078  1.240647  2.266055  2.579587   
2    1.139581  1.592408  1.547358  2.864729  1.606855  3.034000  2.552147   
3    1.115538  1.535857  1.421315  2.940120  1.525948  2.965070  2.567376   
4    1.110347  1.704168  1.895038  2.665340  1.478518  1.966955  2.202553   
..        ...       ...       ...       ...       ...       ...       ...   
292  1.087635  1.769318  1.983138  2.563851  1.552554  1.623851  2.424548   
293  1.101410  1.643631  2.162999  2.422545  1.290014  1.295284  2.576842   
294  1.031936  1.926000  2.153846  2.720665  1.111000  1.319680  2.221053   
295  1.036667  1.824000  1.844000  2.622667  1.150000  1.211333  2.566667   
296  1.038747  1.960363  1.839506  2.352159  1.135091  1.141564  2.391089   

         D059      D060      F115  ...      E027      E028      E033  \
0    2.308442  3.041973  2.148566  ...  2.020095  2.961538  5.069490   
1    2.470076  3.200000  2.347737  ...  1.989822  2.911215  5.340717   
2    3.145473  3.344344  2.878636  ...  1.779899  2.648431  4.638858   
3    3.332997  3.552448  1.888224  ...  2.054835  2.349650  4.857612   
4    2.509280  2.991103  2.311991  ...  2.227924  2.661765  5.261830   
..        ...       ...       ...  ...       ...       ...       ...   
292  2.435270  2.797750  4.398680  ...  2.405430  2.517768  6.287786   
293  2.458937  3.024793  3.481787  ...  2.146834  2.648431  6.683003   
294  2.421444  3.143863  1.553447  ...  2.528184  2.865687  3.610374   
295  2.309333  3.134667  2.340667  ...  2.429333  2.576667  5.224000   
296  2.548037  3.182119  2.099753  ...  2.590191  2.845705  5.431233   

         E035      E036      E037      E039      E040      B008  Country_Wave  
0    4.880041  3.594292  7.062374  3.305414  2.909091  1.570608          ALB3  
1    5.959391  3.848075  6.403015  2.972945  4.350338  1.562570          ALB4  
2    6.143574  4.884694  6.358641  4.423423  5.715145  1.229525          AND5  
3    6.142430  5.062874  6.533400  3.132470  4.575274  1.344546          AND7  
4    5.878989  5.424148  6.211677  3.735065  4.350338  1.524670          ARG1  
..        ...       ...       ...       ...       ...       ...           ...  
292  6.054688  6.238123  5.234501  5.202890  4.756958  1.639196          ZAF6  
293  6.228395  6.069945  6.516107  4.090147  3.686639  1.640969          ZMB5  
294  6.839034  4.271795  7.037074  2.494439  4.350338  1.667010          ZWE4  
295  6.977333  4.742000  7.129333  3.521333  3.615333  1.625333          ZWE6  
296  6.449383  5.512397  7.253707  4.169967  2.497529  1.465174          ZWE7  

[297 rows x 39 columns]

### Reshape data frame

In [5]:
year_dict = {
    '1': [1981, 1982, 1983],
    '2': [1990, 1991, 1992],
    '3': [1995, 1996, 1997, 1998],
    '4': [2000, 2001, 2002, 2003, 2004],
    '5': [2005, 2006, 2007, 2008],
    '6': [2010, 2011, 2012, 2013, 2014],
    '7': [2017, 2018, 2019, 2020, 2021, 2022]
}


In [6]:
df_w_years = df_mean.copy()
df_w_years['Years'] = df_mean['Country_Wave'].str[3].map(year_dict) # this variable 
df_w_years['Country_Wave'] = df_w_years['Country_Wave'].str.extract(r'([A-Za-z]{3})')

Now, the id of each country does not depend on the wave. We add a new variable (Years) that contains the list of years of the wave that the country belongs to.

In [7]:
df_w_years

A001      A002      A003      A004      A005      A006      D057  \
0    1.044266  1.946502  2.573057  3.098765  1.245703  2.260246  2.839045   
1    1.047094  1.855992  2.455375  2.956078  1.240647  2.266055  2.579587   
2    1.139581  1.592408  1.547358  2.864729  1.606855  3.034000  2.552147   
3    1.115538  1.535857  1.421315  2.940120  1.525948  2.965070  2.567376   
4    1.110347  1.704168  1.895038  2.665340  1.478518  1.966955  2.202553   
..        ...       ...       ...       ...       ...       ...       ...   
292  1.087635  1.769318  1.983138  2.563851  1.552554  1.623851  2.424548   
293  1.101410  1.643631  2.162999  2.422545  1.290014  1.295284  2.576842   
294  1.031936  1.926000  2.153846  2.720665  1.111000  1.319680  2.221053   
295  1.036667  1.824000  1.844000  2.622667  1.150000  1.211333  2.566667   
296  1.038747  1.960363  1.839506  2.352159  1.135091  1.141564  2.391089   

         D059      D060      F115  ...      E028      E033      E035  \
0    2.308442  3.041973  2.148566  ...  2.961538  5.069490  4.880041   
1    2.470076  3.200000  2.347737  ...  2.911215  5.340717  5.959391   
2    3.145473  3.344344  2.878636  ...  2.648431  4.638858  6.143574   
3    3.332997  3.552448  1.888224  ...  2.349650  4.857612  6.142430   
4    2.509280  2.991103  2.311991  ...  2.661765  5.261830  5.878989   
..        ...       ...       ...  ...       ...       ...       ...   
292  2.435270  2.797750  4.398680  ...  2.517768  6.287786  6.054688   
293  2.458937  3.024793  3.481787  ...  2.648431  6.683003  6.228395   
294  2.421444  3.143863  1.553447  ...  2.865687  3.610374  6.839034   
295  2.309333  3.134667  2.340667  ...  2.576667  5.224000  6.977333   
296  2.548037  3.182119  2.099753  ...  2.845705  5.431233  6.449383   

         E036      E037      E039      E040      B008  Country_Wave  \
0    3.594292  7.062374  3.305414  2.909091  1.570608           ALB   
1    3.848075  6.403015  2.972945  4.350338  1.562570           ALB   
2    4.884694  6.358641  4.423423  5.715145  1.229525           AND   
3    5.062874  6.533400  3.132470  4.575274  1.344546           AND   
4    5.424148  6.211677  3.735065  4.350338  1.524670           ARG   
..        ...       ...       ...       ...       ...           ...   
292  6.238123  5.234501  5.202890  4.756958  1.639196           ZAF   
293  6.069945  6.516107  4.090147  3.686639  1.640969           ZMB   
294  4.271795  7.037074  2.494439  4.350338  1.667010           ZWE   
295  4.742000  7.129333  3.521333  3.615333  1.625333           ZWE   
296  5.512397  7.253707  4.169967  2.497529  1.465174           ZWE   

                                    Years  
0                [1995, 1996, 1997, 1998]  
1          [2000, 2001, 2002, 2003, 2004]  
2                [2005, 2006, 2007, 2008]  
3    [2017, 2018, 2019, 2020, 2021, 2022]  
4                      [1981, 1982, 1983]  
..                                    ...  
292        [2010, 2011, 2012, 2013, 2014]  
293              [2005, 2006, 2007, 2008]  
294        [2000, 2001, 2002, 2003, 2004]  
295        [2010, 2011, 2012, 2013, 2014]  
296  [2017, 2018, 2019, 2020, 2021, 2022]  

[297 rows x 40 columns]

### Filtering countries by waves

In [8]:
df_w_years

A001      A002      A003      A004      A005      A006      D057  \
0    1.044266  1.946502  2.573057  3.098765  1.245703  2.260246  2.839045   
1    1.047094  1.855992  2.455375  2.956078  1.240647  2.266055  2.579587   
2    1.139581  1.592408  1.547358  2.864729  1.606855  3.034000  2.552147   
3    1.115538  1.535857  1.421315  2.940120  1.525948  2.965070  2.567376   
4    1.110347  1.704168  1.895038  2.665340  1.478518  1.966955  2.202553   
..        ...       ...       ...       ...       ...       ...       ...   
292  1.087635  1.769318  1.983138  2.563851  1.552554  1.623851  2.424548   
293  1.101410  1.643631  2.162999  2.422545  1.290014  1.295284  2.576842   
294  1.031936  1.926000  2.153846  2.720665  1.111000  1.319680  2.221053   
295  1.036667  1.824000  1.844000  2.622667  1.150000  1.211333  2.566667   
296  1.038747  1.960363  1.839506  2.352159  1.135091  1.141564  2.391089   

         D059      D060      F115  ...      E028      E033      E035  \
0    2.308442  3.041973  2.148566  ...  2.961538  5.069490  4.880041   
1    2.470076  3.200000  2.347737  ...  2.911215  5.340717  5.959391   
2    3.145473  3.344344  2.878636  ...  2.648431  4.638858  6.143574   
3    3.332997  3.552448  1.888224  ...  2.349650  4.857612  6.142430   
4    2.509280  2.991103  2.311991  ...  2.661765  5.261830  5.878989   
..        ...       ...       ...  ...       ...       ...       ...   
292  2.435270  2.797750  4.398680  ...  2.517768  6.287786  6.054688   
293  2.458937  3.024793  3.481787  ...  2.648431  6.683003  6.228395   
294  2.421444  3.143863  1.553447  ...  2.865687  3.610374  6.839034   
295  2.309333  3.134667  2.340667  ...  2.576667  5.224000  6.977333   
296  2.548037  3.182119  2.099753  ...  2.845705  5.431233  6.449383   

         E036      E037      E039      E040      B008  Country_Wave  \
0    3.594292  7.062374  3.305414  2.909091  1.570608           ALB   
1    3.848075  6.403015  2.972945  4.350338  1.562570           ALB   
2    4.884694  6.358641  4.423423  5.715145  1.229525           AND   
3    5.062874  6.533400  3.132470  4.575274  1.344546           AND   
4    5.424148  6.211677  3.735065  4.350338  1.524670           ARG   
..        ...       ...       ...       ...       ...           ...   
292  6.238123  5.234501  5.202890  4.756958  1.639196           ZAF   
293  6.069945  6.516107  4.090147  3.686639  1.640969           ZMB   
294  4.271795  7.037074  2.494439  4.350338  1.667010           ZWE   
295  4.742000  7.129333  3.521333  3.615333  1.625333           ZWE   
296  5.512397  7.253707  4.169967  2.497529  1.465174           ZWE   

                                    Years  
0                [1995, 1996, 1997, 1998]  
1          [2000, 2001, 2002, 2003, 2004]  
2                [2005, 2006, 2007, 2008]  
3    [2017, 2018, 2019, 2020, 2021, 2022]  
4                      [1981, 1982, 1983]  
..                                    ...  
292        [2010, 2011, 2012, 2013, 2014]  
293              [2005, 2006, 2007, 2008]  
294        [2000, 2001, 2002, 2003, 2004]  
295        [2010, 2011, 2012, 2013, 2014]  
296  [2017, 2018, 2019, 2020, 2021, 2022]  

[297 rows x 40 columns]

#### Filter 1: drop countries with less than 3 observations

In [9]:
country_counts = df_w_years['Country_Wave'].value_counts()
print("Number of countries: "+str(country_counts.size))
filtered_counts = country_counts[country_counts > 2]
print("Number of filtered countries: "+str(filtered_counts.size))
filtered_counts

Number of countries: 105
Number of filtered countries: 52


Country_Wave
ARG    7
USA    7
MEX    7
KOR    7
JPN    7
CHN    6
CHL    6
ZAF    6
TUR    6
NGA    5
PER    5
AUS    5
BRA    5
CAN    5
IND    5
SWE    5
RUS    5
ESP    5
EGY    4
COL    4
PAK    4
POL    4
PHL    4
NZL    4
ROU    4
IRQ    4
JOR    4
SRB    4
DEU    4
UKR    4
TWN    4
CHE    3
CYP    3
HKG    3
BLR    3
HUN    3
VNM    3
FIN    3
GEO    3
BGD    3
IDN    3
THA    3
PRI    3
SVN    3
MYS    3
KGZ    3
ZWE    3
URY    3
SGP    3
MAR    3
IRN    3
MDA    3
Name: count, dtype: int64

In [10]:
# Extract only the country codes (the index of filtered_counts)
country_codes = filtered_counts.index.tolist()

country_codes

['ARG',
 'USA',
 'MEX',
 'KOR',
 'JPN',
 'CHN',
 'CHL',
 'ZAF',
 'TUR',
 'NGA',
 'PER',
 'AUS',
 'BRA',
 'CAN',
 'IND',
 'SWE',
 'RUS',
 'ESP',
 'EGY',
 'COL',
 'PAK',
 'POL',
 'PHL',
 'NZL',
 'ROU',
 'IRQ',
 'JOR',
 'SRB',
 'DEU',
 'UKR',
 'TWN',
 'CHE',
 'CYP',
 'HKG',
 'BLR',
 'HUN',
 'VNM',
 'FIN',
 'GEO',
 'BGD',
 'IDN',
 'THA',
 'PRI',
 'SVN',
 'MYS',
 'KGZ',
 'ZWE',
 'URY',
 'SGP',
 'MAR',
 'IRN',
 'MDA']

In [11]:
mask_in_codes = df_w_years['Country_Wave'].isin(country_codes)

# Drop the rows where the mask is True
df_w_years_filtered = df_w_years[mask_in_codes]

len(df_w_years_filtered)

219

#### Filter 2: drop observations from waves with few observations

In [12]:
# Extract the wave number from the 'Country_Wave' column
df['Wave'] = df['Country_Wave'].str.extract('(\d+)$')  # Extracts the last digit(s) (wave number)
# Count the number of countries per wave
country_counts_per_wave = df.groupby('Wave')['Country_Wave'].nunique()
total_countries = country_counts_per_wave.sum()
print(total_countries)
country_counts_per_wave


297


Wave
1    11
2    20
3    55
4    41
5    58
6    60
7    52
Name: Country_Wave, dtype: int64

In [13]:
# Create a NumPy array of years that belong to waves 1 and 2 (we want to filter out these waves)
years_wave_1_2 = np.array(year_dict['1'] + year_dict['2'])
# We will apply this second layer of filtering to the df filtered by countries with less than two observations, and we add the Wave col 
df_w_years_filtered_wave = df_w_years_filtered.copy()
df_w_years_filtered_wave['Wave'] = df['Country_Wave'].str.extract('(\d+)$')  # Extracts the last digit(s) (wave number)

# Create a mask for rows where the 'Years' column contains any year in 'years_wave_1_2'
mask = df_w_years_filtered_wave['Years'].apply(lambda x: np.isin(x, years_wave_1_2).any())

# Filter the DataFrame to keep only rows where the mask is False (i.e., not in wave 1 or 2)
df_total_filtered = df_w_years_filtered_wave[~mask]

# Display the filtered DataFrame
print(len(df_total_filtered))
df_total_filtered.head(5)

190


A001      A002      A003      A004      A005      A006      D057  \
6   1.142725  1.686165  2.027128  2.960711  1.340130  2.059813  2.277445   
7   1.118061  1.715190  1.901195  3.134328  1.316286  1.903302  2.061017   
8   1.111111  1.569277  1.780020  2.934542  1.352705  2.160966  2.217241   
9   1.123541  1.570039  1.870117  2.880906  1.580454  2.451581  2.433923   
10  1.091725  1.541376  1.809810  2.807193  1.468405  2.393939  2.387132   

        D059      D060      F115  ...      E033      E035      E036      E037  \
6   2.698364  3.023787  2.468224  ...  5.688073  5.955535  5.075173  5.703810   
7   2.800000  3.110390  2.427330  ...  6.003448  4.899112  5.656000  6.672859   
8   2.829811  3.189979  2.564050  ...  5.526087  5.328542  6.945312  6.088751   
9   2.930470  3.170683  2.531343  ...  5.344994  4.746680  6.212903  6.115822   
10  3.137643  3.253564  3.049645  ...  5.898428  5.565306  6.105263  5.981308   

        E039      E040      B008  Country_Wave  \
6   3.900188  5.051838  1.704612           ARG   
7   4.171913  4.350338  1.710997           ARG   
8   4.711157  4.987705  1.306034           ARG   
9   4.710010  4.216410  1.431203           ARG   
10  4.738730  4.690256  1.583051           ARG   

                                   Years  Wave  
6               [1995, 1996, 1997, 1998]     3  
7         [2000, 2001, 2002, 2003, 2004]     4  
8               [2005, 2006, 2007, 2008]     5  
9         [2010, 2011, 2012, 2013, 2014]     6  
10  [2017, 2018, 2019, 2020, 2021, 2022]     7  

[5 rows x 41 columns]

#### List of countries

In [14]:
df_total_filtered.head(5)

A001      A002      A003      A004      A005      A006      D057  \
6   1.142725  1.686165  2.027128  2.960711  1.340130  2.059813  2.277445   
7   1.118061  1.715190  1.901195  3.134328  1.316286  1.903302  2.061017   
8   1.111111  1.569277  1.780020  2.934542  1.352705  2.160966  2.217241   
9   1.123541  1.570039  1.870117  2.880906  1.580454  2.451581  2.433923   
10  1.091725  1.541376  1.809810  2.807193  1.468405  2.393939  2.387132   

        D059      D060      F115  ...      E033      E035      E036      E037  \
6   2.698364  3.023787  2.468224  ...  5.688073  5.955535  5.075173  5.703810   
7   2.800000  3.110390  2.427330  ...  6.003448  4.899112  5.656000  6.672859   
8   2.829811  3.189979  2.564050  ...  5.526087  5.328542  6.945312  6.088751   
9   2.930470  3.170683  2.531343  ...  5.344994  4.746680  6.212903  6.115822   
10  3.137643  3.253564  3.049645  ...  5.898428  5.565306  6.105263  5.981308   

        E039      E040      B008  Country_Wave  \
6   3.900188  5.051838  1.704612           ARG   
7   4.171913  4.350338  1.710997           ARG   
8   4.711157  4.987705  1.306034           ARG   
9   4.710010  4.216410  1.431203           ARG   
10  4.738730  4.690256  1.583051           ARG   

                                   Years  Wave  
6               [1995, 1996, 1997, 1998]     3  
7         [2000, 2001, 2002, 2003, 2004]     4  
8               [2005, 2006, 2007, 2008]     5  
9         [2010, 2011, 2012, 2013, 2014]     6  
10  [2017, 2018, 2019, 2020, 2021, 2022]     7  

[5 rows x 41 columns]

In [15]:
#Extract unique countries from the "Country_Wave" column
unique_countries = df_total_filtered['Country_Wave'].unique()

# Display the unique countries
print("Unique countries:", unique_countries)

Unique countries: ['ARG' 'AUS' 'BGD' 'BLR' 'BRA' 'CAN' 'CHE' 'CHL' 'CHN' 'COL' 'CYP' 'DEU'
 'EGY' 'ESP' 'FIN' 'GEO' 'HKG' 'HUN' 'IDN' 'IND' 'IRN' 'IRQ' 'JOR' 'JPN'
 'KGZ' 'KOR' 'MAR' 'MDA' 'MEX' 'MYS' 'NGA' 'NZL' 'PAK' 'PER' 'PHL' 'POL'
 'PRI' 'ROU' 'RUS' 'SGP' 'SRB' 'SVN' 'SWE' 'THA' 'TUR' 'TWN' 'UKR' 'URY'
 'USA' 'VNM' 'ZAF' 'ZWE']


### Generate regression variables datasets

#### GDP per capita example

In [24]:
gdp_pc_noisy = pd.read_csv('C:/Users/USERS/MDA_PROJECT/Datasets_Westernization_R_scripts_regression/CleaningWBvar_complete_dataset/GDP_pcapita.csv')
gdp_pc_noisy.tail(10)

Series Name        Series Code  \
46     GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
47     GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
48     GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
49     GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
50     GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
51                                               NaN                NaN   
52                                               NaN                NaN   
53                                               NaN                NaN   
54  Data from database: World Development Indicators                NaN   
55                          Last Updated: 11/13/2024                NaN   

   Country Name Country Code  1995 [YR1995]  1996 [YR1996]  1997 [YR1997]  \
46     Zimbabwe          ZWE    1977.848035    2186.114786    2246.401152   
47     Thailand          THA    6862.812134    7287.021553    7116.612033   
48      Turkiye          TUR   10249.163562   10969.122749   11785.239427   
49      Ukraine          UKR    4136.524414    3824.812500    3808.598145   
50      Uruguay          URY    8853.266288    9464.140712   10390.889452   
51          NaN          NaN            NaN            NaN            NaN   
52          NaN          NaN            NaN            NaN            NaN   
53          NaN          NaN            NaN            NaN            NaN   
54          NaN          NaN            NaN            NaN            NaN   
55          NaN          NaN            NaN            NaN            NaN   

    1998 [YR1998]  1999 [YR1999]  2000 [YR2000]  ...  2013 [YR2013]  \
46    2299.553828    2279.838655    2237.579265  ...    2558.994062   
47    6565.803550    6885.400083    7282.310401  ...   15274.232012   
48    8969.162706    8649.614773    9498.964414  ...   22373.246558   
49    3812.138672    3894.897461    4260.801758  ...   11056.541992   
50   10923.208717   10855.131269   10842.927482  ...   20442.353492   
51            NaN            NaN            NaN  ...            NaN   
52            NaN            NaN            NaN  ...            NaN   
53            NaN            NaN            NaN  ...            NaN   
54            NaN            NaN            NaN  ...            NaN   
55            NaN            NaN            NaN  ...            NaN   

    2014 [YR2014]  2015 [YR2015]  2016 [YR2016]  2017 [YR2017]  2018 [YR2018]  \
46    2618.240859    2692.290700    2825.357165    2435.794358    2611.326409   
47   15421.415567   15846.523158   16730.941042   17638.277308   18942.109761   
48   24105.023798   25855.765356   26695.915690   28193.174485   28299.398872   
49   10664.624023   10064.842773   11011.970703   11686.857422   12708.798828   
50   21640.033213   21941.890552   22609.834987   23373.752518   24155.108846   
51            NaN            NaN            NaN            NaN            NaN   
52            NaN            NaN            NaN            NaN            NaN   
53            NaN            NaN            NaN            NaN            NaN   
54            NaN            NaN            NaN            NaN            NaN   
55            NaN            NaN            NaN            NaN            NaN   

    2019 [YR2019]  2020 [YR2020]  2021 [YR2021]  2022 [YR2022]  
46    2711.529334    2784.872395    3276.267753    3660.835501  
47   20023.445713   19208.200446   20280.933715   22213.799375  
48   28461.179032   28680.200941   31637.530544   38355.153967  
49   14380.861328   15716.841797   18040.345703   16080.228516  
50   25548.063384   25551.337054   29187.428675   32746.315276  
51            NaN            NaN            NaN            NaN  
52            NaN            NaN            NaN            NaN  
53            NaN            NaN            NaN            NaN  
54            NaN            NaN            NaN            NaN  
55            NaN            NaN           

In [25]:
gdp_pc = gdp_pc_noisy.dropna(subset=['Country Code'])
gdp_pc.tail(10)

Series Name        Series Code  \
41  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
42  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
43  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
44  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
45  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
46  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
47  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
48  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
49  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   
50  GDP per capita, PPP (current international $)  NY.GDP.PCAP.PP.CD   

     Country Name Country Code  1995 [YR1995]  1996 [YR1996]  1997 [YR1997]  \
41       Slovenia          SVN   13618.504665   14303.317376   15279.962238   
42         Sweden          SWE   23090.716896   23925.739694   24801.241684   
43  United States          USA   28690.875701   29967.712718   31459.129969   
44       Viet Nam          VNM    1798.681903    1970.956874    2136.682863   
45   South Africa          ZAF    6962.010411    7282.474521    7496.027146   
46       Zimbabwe          ZWE    1977.848035    2186.114786    2246.401152   
47       Thailand          THA    6862.812134    7287.021553    7116.612033   
48        Turkiye          TUR   10249.163562   10969.122749   11785.239427   
49        Ukraine          UKR    4136.524414    3824.812500    3808.598145   
50        Uruguay          URY    8853.266288    9464.140712   10390.889452   

    1998 [YR1998]  1999 [YR1999]  2000 [YR2000]  ...  2013 [YR2013]  \
41   16055.950900   17083.251722   18003.519862  ...   29973.699354   
42   25887.656333   27471.558272   29622.038475  ...   46312.392382   
43   32853.672595   34515.381307   36329.970260  ...   53409.750777   
44    2253.550810    2364.032066    2552.997526  ...    6781.267213   
45    7523.949149    7727.179545    8155.205490  ...   13569.252848   
46    2299.553828    2279.838655    2237.579265  ...    2558.994062   
47    6565.803550    6885.400083    7282.310401  ...   15274.232012   
48    8969.162706    8649.614773    9498.964414  ...   22373.246558   
49    3812.138672    3894.897461    4260.801758  ...   11056.541992   
50   10923.208717   10855.131269   10842.927482  ...   20442.353492   

    2014 [YR2014]  2015 [YR2015]  2016 [YR2016]  2017 [YR2017]  2018 [YR2018]  \
41   30870.023667   31628.247176   33936.044064   36507.553042   38971.824116   
42   47184.669134   49103.133434   50430.252820   51947.954248   53521.630643   
43   55304.315532   57040.208214   58206.614193   60322.261424   63201.045848   
44    7332.395857    7724.161017    8453.065219    9281.283364   10149.498164   
45   13570.149711   13600.616798   13719.943192   13979.523875   13643.774139   
46    2618.240859    2692.290700    2825.357165    2435.794358    2611.326409   
47   15421.415567   15846.523158   16730.941042   17638.277308   18942.109761   
48   24105.023798   25855.765356   26695.915690   28193.174485   28299.398872   
49   10664.624023   10064.842773   11011.970703   11686.857422   12708.798828   
50   21640.033213   21941.890552   22609.834987   23373.752518   24155.108846   

    2019 [YR2019]  2020 [YR2020]  2021 [YR2021]  2022 [YR2022]  
41   42747.962112   42040.909061   46502.097894   51281.857510  
42   57229.776747   57690.324459   63375.044646   68088.193267  
43   65548.070785   64317.398913   71055.876194   77246.673883  
44   11190.153191   11781.148298   12230.259622   14051.248771  
45   13706.713816   13062.672773   14172.544251   15330.769274  
46    2711.529334    2784.872395    3276.267753    3660.835501  
47   20023.445713   19208.200446   20280.933715   22213.799375  
48   28461.179032   28680.200941   31637.530544   38355.153967  
49   14380.861328   15716.841797   18040.345703   16080.228516  
50   25548.063384   25551.337054   29187.428675   3

In [26]:
# Remove `[YR...]` from the year columns
gdp_pc.columns = [col.split()[0] if "[YR" in col else col for col in gdp_pc.columns]

# Esure that the column names of years are integers
gdp_pc.columns = [int(col) if col.isdigit() else col for col in gdp_pc.columns] 
column_name_types2 = [type(col) for col in gdp_pc.columns]
print(column_name_types2)

# Drop the first three columns: 'Series Name', 'Series Code', 'Country Name'
gdp_pc = gdp_pc.iloc[:, 3:]

# Rename the fourth column to "Country_Wave"
gdp_pc = gdp_pc.rename(columns={"Country Code": "Country_Wave"})

# Display the cleaned DataFrame
gdp_pc.head(3)

[<class 'str'>, <class 'str'>, <class 'str'>, <class 'str'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>, <class 'int'>]


Country_Wave          1995          1996          1997          1998  \
0          ARG   9925.995407  10532.743210  11445.451086  11881.298960   
1          AUS  21038.666785  22132.192903  23124.963912  24378.245311   
2          BGD   1255.597114   1313.185822   1371.076758   1431.160469   

           1999          2000          2001          2002          2003  ...  \
0  11508.157046  11544.408971  11160.567802   9991.382124  10975.128236  ...   
1  25485.391446  26541.665321  27645.814029  29032.490956  30121.818418  ...   
2   1490.379121   1574.525647   1659.895464   1718.083880   1803.270265  ...   

           2013          2014          2015          2016          2017  \
0  20108.205701  19649.352891  20058.338639  20248.721387  23514.667087   
1  45936.049310  46914.386708  46292.095439  47289.285914  48418.558422   
2   3751.061086   4023.541146   4253.687699   4608.574652   4894.442197   

           2018          2019          2020          2021          2022  
0  24498.378927  23535.012751  22302.150489  26086.802355  29037.489122  
1  50251.335338  52746.718290  54064.079466  57406.178760  65365.948511  
2   5484.763547   6025.087349   6596.044598   7366.489971   8354.715868  

[3 rows x 29 columns]

In [27]:
# Ensure that the country codes for this dataset are the same as for our previous data
country_wave_set_gdp = set(gdp_pc['Country_Wave'])

# Convert unique_countries to a set for comparison
#unique_countries_set = set(unique_countries)

# Check if both sets are the same
if country_wave_set_gdp == unique_countries_set:
    print("The elements in 'unique_countries' match exactly with 'Country_Wave'!")
else :
    print("There are mismatches")
    list_1 = list(country_wave_set_gdp - unique_countries_set)  # Elements in Country_Wave but not in unique_countries
    list_2 = list(unique_countries_set - country_wave_set_gdp)  # Elements in unique_countries but not in Country_Wave
    print(list_1)
    print(list_2)

There are mismatches
['NIC']
['TWN', 'NGA']


In [28]:
# Update the unmatched names
gdp_pc['Country_Wave'] = gdp_pc['Country_Wave'].replace('NIC', 'NGA')

In [29]:
gdp_pc_nan = gdp_pc.copy()
gdp_pc_nan.replace("..", np.nan, inplace=True)

#Check if missing values were correcly replaced by NaNs
print(gdp_pc_nan.isna().sum().sum())  # Total number of NaN values in the dataset

0


In [30]:
# Export as csv
#gdp_pc_nan.to_csv('gdp_pc.csv', index=False)

### Computing the mean of the variables

 Variables: independent variable (crime rate) and variables of interest (GDP, unemployment), etc

In [38]:
import numpy as np
import pandas as pd

years = list(range(1980, 2023))

# Generate random numbers for two rows (2 rows, 43 columns for years 1980 to 2022)
data = np.random.rand(3, len(years))  # 2 rows, 43 columns

# Create a DataFrame with 'Country_Wave' values as 'ALB' and 'A'
GDP = pd.DataFrame(data, columns=years)
GDP['Country_Wave'] = ['ALB', 'AND', 'ARG']  # Two entries for Country_Wave

# Reorder the columns so 'Country_Wave' is the first column
GDP = GDP[['Country_Wave'] + years]

In [39]:
GDP

Country_Wave      1980      1981      1982      1983      1984      1985  \
0          ALB  0.825310  0.869857  0.849944  0.944724  0.431707  0.344559   
1          AND  0.049895  0.064664  0.291515  0.113644  0.168910  0.193293   
2          ARG  0.773209  0.391257  0.810463  0.742187  0.049113  0.972791   

       1986      1987      1988  ...      2013      2014      2015      2016  \
0  0.078293  0.825811  0.670429  ...  0.584984  0.811456  0.287432  0.600514   
1  0.540854  0.361594  0.456203  ...  0.862401  0.296187  0.100170  0.254995   
2  0.411344  0.823423  0.505344  ...  0.973189  0.596173  0.637157  0.004698   

       2017      2018      2019      2020      2021      2022  
0  0.666708  0.867674  0.171858  0.536739  0.392037  0.279032  
1  0.972016  0.146568  0.241099  0.291901  0.352152  0.605260  
2  0.681752  0.609561  0.860732  0.362028  0.733346  0.872876  

[3 rows x 44 columns]

In [85]:
def calculate_mean(row, df, label):
    # Extract the years list from the row. Row = row from the df_w_years dataset (the one with the questions)
    years_list = row[label]
    years_list = [str(year) for year in years_list]

    df = df[df['Country_Wave'] == row['Country_Wave']]
    df[years_list].fillna(-1)

    values = df[years_list].values.flatten()
    
    values = values[values > 0]
    
    # Calculate and return the mean of those values
    return values.mean() if len(values) > 0 else None

In [86]:
# Load all datasets

crimeRate = pd.read_csv('C:/Users/USERS/MDA_PROJECT/World_Bank_Variables/cr_nan.csv')
gdpGrowth = pd.read_csv('C:/Users/USERS/MDA_PROJECT/World_Bank_Variables/Growth_rate.csv')
gdpPc = pd.read_csv('C:/Users/USERS/MDA_PROJECT/World_Bank_Variables/gdp_pc.csv')
unempRate = pd.read_csv('C:/Users/USERS/MDA_PROJECT/World_Bank_Variables/unemp.csv')
education = pd.read_csv('C:/Users/USERS/MDA_PROJECT/World_Bank_Variables/Education.csv')

In [87]:
df_filtered = df_w_years[df_w_years['Country_Wave'].isin(crimeRate['Country_Wave'])]
remove_years = [[1981, 1982, 1983], [1990, 1991, 1992]]
df_filtered_cleaned = df_filtered[~df_filtered['Years'].apply(lambda x: x in remove_years)]

In [88]:
label='Years'

# Add Crime rate
df_cr = df_filtered2.copy()
df_cr['Crime_Rate'] = df_cr.apply(calculate_mean, axis=1, df=crimeRate, label=label)

# Add GDP growth
df_gdpG = df_cr.copy()
df_gdpG['GDP_growth'] =df_gdpG.apply(calculate_mean, axis=1, df=gdpGrowth, label=label)

# Add GDP per capita
df_gdpPc = df_gdpG.copy()
df_gdpPc['GDP_pc'] = df_gdpPc.apply(calculate_mean, axis=1, df=gdpPc, label=label)

# Add Unemployment rate
df_unemp = df_gdpPc.copy()
df_unemp['Unemp_Rate'] = df_unemp.apply(calculate_mean, axis=1, df=unempRate, label=label)

# Add Education
df_ed = df_unemp .copy()
df_ed['Education'] = df_ed.apply(calculate_mean, axis=1, df=education, label=label)

In [96]:
df_final = df_ed.copy()
df_final.head(2)

A001      A002      A003      A004      A005      A006      D057  \
6  1.142725  1.686165  2.027128  2.960711  1.340130  2.059813  2.277445   
7  1.118061  1.715190  1.901195  3.134328  1.316286  1.903302  2.061017   

       D059      D060      F115  ...      E039      E040      B008  \
6  2.698364  3.023787  2.468224  ...  3.900188  5.051838  1.704612   
7  2.800000  3.110390  2.427330  ...  4.171913  4.350338  1.710997   

   Country_Wave                           Years  Crime_Rate  GDP_growth  \
6           ARG        [1995, 1996, 1997, 1998]         NaN    5.829305   
7           ARG  [2000, 2001, 2002, 2003, 2004]    7.892241    8.933307   

         GDP_pc  Unemp_Rate  Education  
6  10946.372166     15.8450   11.64263  
7  11167.054004     16.1584   12.94255  

[2 rows x 45 columns]

In [97]:
# Reverse mapping from year list to year ID
year_to_id = {tuple(value): key for key, value in year_dict.items()}

# Add digit to Country_Wave based on the Years column
df_final['Country_Wave'] = df_final['Country_Wave'] + df_final['Years'].apply(lambda x: year_to_id[tuple(x)])
df_final.head(5)

A001      A002      A003      A004      A005      A006      D057  \
6   1.142725  1.686165  2.027128  2.960711  1.340130  2.059813  2.277445   
7   1.118061  1.715190  1.901195  3.134328  1.316286  1.903302  2.061017   
8   1.111111  1.569277  1.780020  2.934542  1.352705  2.160966  2.217241   
9   1.123541  1.570039  1.870117  2.880906  1.580454  2.451581  2.433923   
10  1.091725  1.541376  1.809810  2.807193  1.468405  2.393939  2.387132   

        D059      D060      F115  ...      E039      E040      B008  \
6   2.698364  3.023787  2.468224  ...  3.900188  5.051838  1.704612   
7   2.800000  3.110390  2.427330  ...  4.171913  4.350338  1.710997   
8   2.829811  3.189979  2.564050  ...  4.711157  4.987705  1.306034   
9   2.930470  3.170683  2.531343  ...  4.710010  4.216410  1.431203   
10  3.137643  3.253564  3.049645  ...  4.738730  4.690256  1.583051   

    Country_Wave                                 Years  Crime_Rate  \
6           ARG3              [1995, 1996, 1997, 1998]         NaN   
7           ARG4        [2000, 2001, 2002, 2003, 2004]    7.892241   
8           ARG5              [2005, 2006, 2007, 2008]    5.556121   
9           ARG6        [2010, 2011, 2012, 2013, 2014]    6.589016   
10          ARG7  [2017, 2018, 2019, 2020, 2021, 2022]    5.154484   

    GDP_growth        GDP_pc  Unemp_Rate  Education  
6     5.829305  10946.372166   15.845000  11.642630  
7     8.933307  11167.054004   16.158400  12.942550  
8     7.490924  15566.961752    9.472750  12.699400  
9     6.178225  19355.236577    7.295800  13.173304  
10    6.164294  24829.083455    9.068833  13.422358  

[5 rows x 45 columns]

In [99]:
df_final.to_csv('complete_dataset.csv', index=False)

In [100]:
# Total number of NaN values in the dataset
print(df_final.isna().sum().sum())  

47


In [101]:
print(len(df_final))

186
